File: packages/serve-sim/src/device-pose.ts
/** Named fold poses. 0° is closed (cover), 180° is flat open (inner). */
export type DevicePoseId = "closed" | "open" | "book" | "tent" | "tabletop";

export type DevicePoseSpec = {
  id: DevicePoseId;
  label: string;
  hingeDegrees: number;
  coverActive: boolean;
  width: number;
  height: number;
};

export const DUO_COVER_SIZE = { width: 1398, height: 2034 } as const;
export const DUO_INNER_SIZE = { width: 2007, height: 2853 } as const;
export const DUO_COVER_ACTIVE_BELOW_DEGREES = 90;

const PRESETS: readonly DevicePoseSpec[] = (
  [
    ["closed", "Closed", 0],
    ["tent", "Tent", 80],
    ["tabletop", "Table", 100],
    ["book", "Book", 130],
    ["open", "Open", 180],
  ] as const
).map(([id, label, hingeDegrees]) => spec(id, label, hingeDegrees));

function spec(id: DevicePoseId, label: string, hingeDegrees: number): DevicePoseSpec {
  const coverActive = hingeDegrees < DUO_COVER_ACTIVE_BELOW_DEGREES;
  const size = coverActive ? DUO_COVER_SIZE : DUO_INNER_SIZE;
  return { id, label, hingeDegrees, coverActive, width: size.width, height: size.height };
}

export function devicePosePresets(): readonly DevicePoseSpec[] {
  return PRESETS;
}

export function resolveDevicePose(raw: string): DevicePoseSpec | null {
  const name = raw.trim().toLowerCase();
  if (name === "cover") return resolveDevicePose("closed");
  if (name === "inner" || name === "openflat" || name === "flat") return resolveDevicePose("open");
  return PRESETS.find((pose) => pose.id === name) ?? null;
}

export function poseForDisplayRole(role: string): DevicePoseSpec | null {
  if (role === "cover") return resolveDevicePose("closed");
  if (role === "inner") return resolveDevicePose("open");
  return null;
}

export type DuoPanel = "cover" | "inner";

/**
 * The live panel. SpringBoard's primary-panel readback wins. The 90° hinge
 * rule applies only before that readback exists.
 */
export function duoPanel(input: { primaryPanel?: DuoPanel | null; hingeDegrees?: number | null }): DuoPanel {
  if (input.primaryPanel) return input.primaryPanel;
  return (input.hingeDegrees ?? 0) < DUO_COVER_ACTIVE_BELOW_DEGREES ? "cover" : "inner";
}

export function displaySizeForPanel(panel: DuoPanel): {
  coverActive: boolean;
  width: number;
  height: number;
} {
  const coverActive = panel === "cover";
  const size = coverActive ? DUO_COVER_SIZE : DUO_INNER_SIZE;
  return { coverActive, width: size.width, height: size.height };
}

/** Cover vs inner framebuffer size for a raw hinge angle in degrees. */
export function displaySizeForHingeDegrees(hingeDegrees: number): {
  coverActive: boolean;
  width: number;
  height: number;
} {
  return displaySizeForPanel(duoPanel({ hingeDegrees }));
}

File: packages/serve-sim/Sources/SimNativeSupport/DevicePose.swift
/// Named hinge angles for iPhone Duo (V68).
///
/// Which framebuffer is live is decided in TypeScript from SpringBoard's
/// primary-panel readback, with the 90° rule only before that readback.
/// Touch routing uses the captured pixel size (`touchTarget`), not the pose.
/// Do not toggle SimScreen power — FrontBoard's bookend display monitor
/// asserts if a foldable panel is missing at SpringBoard launch.
public struct DevicePoseSpec: Equatable, Sendable {
    public let name: String
    public let hingeDegrees: Double

    public init(name: String, hingeDegrees: Double) {
        self.name = name
        self.hingeDegrees = hingeDegrees
    }
}

public enum DevicePose {
    public static let coverWidth = 1398
    public static let coverHeight = 2034
    public static let innerWidth = 2007
    public static let innerHeight = 2853

    /// Duo's own Integrated screen registrations, including SimulatorHID's mask.
    /// Do not apply this mapping to TVOut, CarPlay, or single-panel devices.
    public static func touchTarget(width: Int, height: Int) -> UInt32? {
        let size = [width, height].sorted()
        if size == [coverWidth, coverHeight] { return 0x40000001 }
        if size == [innerWidth, innerHeight] { return 0x40000003 }
        return nil
    }

    public static let presets: [DevicePoseSpec] = [
        DevicePoseSpec(name: "closed", hingeDegrees: 0),
        DevicePoseSpec(name: "tent", hingeDegrees: 80),
        DevicePoseSpec(name: "tabletop", hingeDegrees: 100),
        DevicePoseSpec(name: "book", hingeDegrees: 130),
        DevicePoseSpec(name: "open", hingeDegrees: 180),
    ]

    public static func preset(named raw: String) -> DevicePoseSpec? {
        let name = raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if name == "cover" { return preset(named: "closed") }
        if name == "inner" || name == "openflat" || name == "flat" {
            return preset(named: "open")
        }
        return presets.first { $0.name == name }
    }
}
